### ISSUE TITLE
eth_call with a specified block that has not been fetched by the node yet return zero values

### ISSUE THREAD

**Issue by wighawag** [ethereum/go-ethereum#15102]
#### Expected behaviour
eth_call should return a failure if the block requested is not present on the node
This way we can disregard the data when such error occur.

**Else you are never sure that your data is actually zero at that particular block or it is the node that has not synced fully yet.**

#### Actual behaviour
Currently eth_call will return 0x for calls with a specified block number that the node does not have

this means that with the following contract:

```
pragma solidity 0.4.11;

contract BlockTest {
     function getBlockInfo() constant returns (
        uint256 blockNumber, 
        uint256 timestamp, 
        bytes32 previousBlockHash
        ){
        timestamp = block.timestamp;
        blockNumber = block.number;
        previousBlockHash = block.blockhash(blockNumber-1);
    }
}
```

or this one :
```
pragma solidity 0.4.11;

contract BlockTest2 {
    uint256 sValue = 7;
     function getBlockInfo() constant returns (
        uint256 storageValue,
        uint256 constantValue,
        uint256 blockNumber, 
        uint256 timestamp, 
        bytes32 previousBlockHash
        ){
        storageValue = sValue;
        constantValue=4;
        timestamp = block.timestamp;
        blockNumber = block.number;
        previousBlockHash = block.blockhash(blockNumber-1);
    }
}

```

an eth_call to getBlockInfo will return 0 for blockNumber, timestamp, previousBlockHash, or any data actually, even constant values.

Of course in the particular example we could use the values to detect such missing block states since these should never be zero but there are many case where we do not need nor want to return these values

#### Steps to reproduce the behaviour
- compile and deploy the contract above 
(on ropsten BlockTest is at : 0xe315fe71cbe04f3dcb5daeeead5928c1b9ee8486  )  
(on ropsten BlockTest2 is at : 0x50c6adc6370847bc00fa365136dd1d45736c47bc  )  

- do an eth_call as follow :
```
{"id":601555241713,"jsonrpc":"2.0","method":"eth_call","params":[{"to":"0xe315fe71cbe04f3dcb5daeeead5928c1b9ee8486","data":"0x00819439"},"0x1060EEF"]}
```
where "0x1060EEF" is a future block

you get :
```
{
    "jsonrpc": "2.0",
    "id": 601555241713,
    "result": "0x"
}
```

while you should get an jsonrpc error

#### Notes

This behaviour have been tested on geth but it might be present on other implementation. If it is the case, should we discuss the issue on the EIP repo ?

Thanks

**Comment by wighawag** [comment]
I added to the test another contract

```
pragma solidity 0.4.11;

contract BlockTest3 {
    uint256 sValue = 0;
     function getBlockInfo() constant returns (
        uint256 storageValue,
        uint256 constantValue,
        uint256 blockNumber, 
        uint256 timestamp, 
        bytes32 previousBlockHash
        ){
        storageValue = sValue;
        constantValue=0;
        timestamp = 0;
        blockNumber = 0;
        previousBlockHash = 0;
    }
}

```
ropsten address : 0xb1bdafcde20ab7bcb61d194069a0aa5db5c7c526

and with a call to an existing block I get
```
 {
        "jsonrpc": "2.0",
        "id": 365692253423,
        "result": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
    }
```
while if the blockNumber was not exisiting yet I get "0x" as result only

I still think an error is better suited but if we stay with "0x" then the issue is in web3 since it does not currently detect such case and will give zero values to every result parameters.

**Comment by wighawag** [comment]
I realized that "0x" is also the value returned when calling a method on block where the contract has not been deployed yet.

I think that in this case "0x" is valid since that the behavior to expect when there is no such method.

In the case where the block has not been fetch yet, on the other hand, this is misleading since if it will be  executed at a later time (when the block is finally fetched) it would give a different answer. Something not possible otherwise in the world of blockchain where everything is supposed to be immutable.

**Comment by wighawag** [comment]
In the context of a node service like Infura where nodes are not necessarily (rarely when under pressure) on the same block, a eth_call on that service (with a specified block) will sometime return "0x" (when not yet synced up to that block) and sometime some actual data

**Comment by wighawag** [comment]
I just tried with a partially synced parity node using default command line param except for the ropsten network:
 
I get this both for a future block (0x1060EEF) and a past block (0xF)

```
{
    "jsonrpc": "2.0",
    "error": {
        "code": -32000,
        "message": "This request is not supported because your node is running with state pruning. Run with --pruning=archive."
    },
    "id": 365692253423
}
```

upon restarting with --pruning=archive (not deleting any data though but it seems to be syncing from scratch)

I get

```
    {
        "jsonrpc": "2.0",
        "result": "0x",
        "id": 365692253423
    }
```

for a past block ("0xF" for example) which did not contain code for the contract address yet

but I get :

```
{
    "jsonrpc": "2.0",
    "error": {
        "code": -32000,
        "message": "This request is not supported because your node is running with state pruning. Run with --pruning=archive."
    },
    "id": 365692253423
}
```
for a block not yet synced (to be clear this is running with parity --pruning=archive)

Could not try it with the node synced up a block past the time the contract was deployed (0x18BD6A) yet but I expect it to give the proper data values since otherwise there will be no way to get any data

In other words it seems parity behave fine

**Comment by wighawag** [comment]
I managed to test on a fully synced node and I get the proper result for a existing block

For a future block (0x1060EEF for example) I get the error message

So I can confirm parity behave as I would expect, except for the message that can be misleading

**Comment by wighawag** [comment]
I had a look at the code and I believe the issue is here (for eth_call) : https://github.com/ethereum/go-ethereum/blob/c91f7beb53ff1ab0376d9aa5fab5a8de8b04631c/internal/ethapi/api.go#L587

When not finding a block, it act if  there were no error (since err will be nil too here)

A fix could be done there : https://github.com/ethereum/go-ethereum/blob/c4d21bc8e548d909d5b2940cb00d3d068f5697ec/eth/api_backend.go#L92 

by only checking for err != null here and adding a check for header == nil where in that case an error will be returned for the missing block 

so every call that specific a block Number would get the proper behaviour

**Comment by egalano** [comment]
This also seems to be affecting nodes that are under heavy RPC load. Our nodes are processing tens of thousands of requests per second and during large traffic spikes some eth_calls that should be returning non-zero balances or values are returning zero. @Arachnid  or @karalabe could you please verify @wighawag's suggested fix?

**Comment by wighawag** [comment]
Hi @egalano 
There is a PR for it here : https://github.com/ethereum/go-ethereum/pull/3485 but it has not been accepted yet. This is a rather urgent and simple thing to fix in my opinion. geth rpc api currently misbehave.
